cli/src/commands/run.rs

```diff
@@ -14,12 +14,262 @@
+    commit: Commit,
+    /// The output directory of the commit, contains stdout and stderr for it
+    output_dir: PathBuf,
+    /// Self-explanatory
+    working_copy_dir: PathBuf,
+    state_dir: PathBuf,
+    /// The `stdout` of the commit
+    stdout: File,
+    /// The `stderr` of the commit
+    stderr: File,
+}
+
+impl StoredCommit {
+    fn new(
+        name: Option<String>,
+        commit: &Commit,
+        output_dir: PathBuf,
+        working_copy_dir: PathBuf,
+        state_dir: PathBuf,
+        stdout: File,
+        stderr: File,
+    ) -> Self {
+        Self {
+            name,
+            commit: commit.clone(),
+            output_dir,
+            working_copy_dir,
+            state_dir,
+            stdout,
+            stderr,
+        }
+    }
+}
+
+fn create_output_files(id: CommitId, path: &PathBuf) -> Result<(File, File), io::Error> {
+    let stdout_path = path.join("output").join(format!("stdout.{id}"));
+    let stderr_path = path.join("output").join(format!("stderr.{id}"));
+    let mut file_options = OpenOptions::new();
+    let stdout = file_options.write(true).open(stdout_path)?;
+    let stderr = file_options.write(true).open(stderr_path)?;
+    Ok((stdout, stderr))
+}
+
+fn create_working_copies(
+    repo_path: &Path,
+    commits: &[Commit],
+) -> Result<Vec<StoredCommit>, io::Error> {
+    let mut results = vec![];
+    // TODO: should be stored in a backend and not hardcoded.
+    let base_path = repo_path.join("run").join("default");
+    for commit in commits {
+        let name = to_wc_name(commit.tree_id());
+
+        let (output_dir, working_copy_dir, state_dir) = create_working_copy_paths(&base_path)?;
+        let (stdout, stderr) = create_output_files(commit.id(), &base_path)?;
+
+        let stored_commit = StoredCommit::new(
+            Some(name),
+            commit,
+            output_dir,
+            working_copy_dir,
+            state_dir,
+            stdout,
+            stderr,
+        );
+        results.push(stored_commit);
+    }
+    Ok(results)
+}
+
+/// The result of a single command invocation in `run_inner`.
+struct RunJobResult {
+    /// The old `CommitId` of the commit.
+    old_id: CommitId,
+    /// The new `CommitId` for the commit.
+    rewritten_id: CommitId,
+    /// The new tree generated from the commit.
+    new_tree: Tree,
+}
+
+// TODO: make this more revset stream friendly.
+async fn run_inner<'a>(
+    tx: WorkspaceCommandTransaction<'a>,
+    sender: Sender<RunJobResult>,
+    jobs: usize,
+    shell_command: &str,
+    commits: &[StoredCommit],
+) -> Result<(), RunError> {
+    let mut command_futures = JoinSet::new();
+    for commit in commits {
+        command_futures.spawn(rewrite_commit(tx, commit.deref(), shell_command));
+    }
+
+    while let Some(res) = command_futures.join_next().await {
+        let done = res?;
+        sender.send(done?);
+    }
+    Ok(())
+}
+
+/// Rewrite a single `StoredCommit`.
+async fn rewrite_commit<'a>(
```

Comment:
issue: This doesn't actually rewrite any commits, so it shouldn't be called `rewrite_commit`.

Reply:
it does rewrite the commits tree and I want to rewrite it in there if possible.

Reply:
imo, it does the core part now, so the name could stay.